### ISSUE TITLE
[QUESTION] How to enable freetype on Windows?

### ISSUE THREAD

**Issue by Ziriax** [mono/SkiaSharp#1505]
It seems the Windows version of SkiaSharp is hardcoded to use DirectWrite?

I would like to be able to configure it (at runtime) to use Freetype instead.

Is this possible? Would a PR be accepted for doing so?

**Comment by Ziriax** [comment]
Interestingly, internally Skia code allows overriding the font manager for unit testing:

```
// A global function pointer that's not declared, but can be overriden at startup by test tools.
sk_sp<SkFontMgr> (*gSkFontMgr_DefaultFactory)() = nullptr;

sk_sp<SkFontMgr> SkFontMgr::RefDefault() {
    static SkOnce once;
    static sk_sp<SkFontMgr> singleton;

    once([]{
        sk_sp<SkFontMgr> fm = gSkFontMgr_DefaultFactory ? gSkFontMgr_DefaultFactory()
                                                        : SkFontMgr::Factory();
        singleton = fm ? std::move(fm) : sk_make_sp<SkEmptyFontMgr>();
    });
    return singleton;
}
```

So in theory, this could be used to configure the font manager at runtime, at least once, at startup?

**Comment by Gillibald** [comment]
I think there should be a proxy font manager that can be used to expose all needed functions as pointers. That would allow us to have a managed font manager and also allows us to chose an implementation at runtime.

**Comment by Ziriax** [comment]
It required very few changes to allow freetype on Windows, switchable once, at runtime.

Not sure if this is something typical SkiaSharp users need though...

However, since it is optional, and doesn't change existing behavior, it certainly is a nice to have? 

It allows the same rendering on all platforms...

**Comment by Ziriax** [comment]
Oh my, in my own tests, FreeType gives better quality rendering than DirectWrite on Windows! Crazy... So yet another reason why to allow enabling it 😉

**Comment by Gillibald** [comment]
Using Freetype in all platforms allows us the have proper render tests that produce the same results on every platform. I guess we just need to add a new build target for Windows that enables Freetype.

**Comment by Ziriax** [comment]
Yes, this is indeed also a use case for us.

Actually, for comparing the images in our unit tests, I ported the excellent PixelMatch library to .NET Standard:

https://github.com/Ziriax/PixelMatch.net

Note that due to CPU architecture differences, you rarely end up with exactly the same rendering when running the code on different machines...

**Comment by mattleibow** [comment]
@Ziriax could you open a PR that switches to freetype? I want to see what is needed and if all the tests are still good.

There are ways to do this, the easiest might be to have a separate package for now that allows a different libSkiaSharp. I do this for the NanoServer package. At least then folks can use this until a nice runtime check can be implemented.

**Comment by Ziriax** [comment]
A separate maintained package is perfect!

Yes, I can provide a PR. We have two options here:

1. make a truly platform independent version of libSkiaSharp, that uses the same technology on all platforms, and *excludes* platform dependent code. It would also exclude any "system global" behavior, e.g. never load system installed fonts, always require explicit typeface construction from a stream/file, etc... Skia's `SkFontMgr_New_Custom_Empty` does this (it uses FreeType). We might not get this right the first time, this will be an incremental effort. This only requires changes to the `.gn` files.

2. allow turning on the custom empty font manager at runtime, but this must be the first call before any other Skia calls are made. This would require changes to both the `C#`, `cpp` and `.gn` files. I personally don't need this. The only reason I mentioned this, is because it would allow for backwards compatibility in the main SkiaSharp package. But it will make the package bigger by dragging in all the FreeType code, and people might complain.

My personal preference goes to (1): a truly cross platform unit testable rendering library (give or take a `bit` because of CPU floating point rounding differences :wink:).

**Comment by mattleibow** [comment]
Option 1 sounds good.

If you can make the or add a variant or some args in the cake, then that should do it very well. Exactly like we do with Linux. In fact, nano server just calls into the main windows cake, so that might be nice if changes are made.

**Comment by soerendd** [comment]
I also did this once, but at least for free type on Linux you need smth like fontconfig. This means on windows we need smth similar

**Comment by Ziriax** [comment]
I our case we didn't need fontconfig, since we load the type faces manually. At least I got it working without fontconfig as far as I recall.

**Comment by soerendd** [comment]
Yes, that's exactly what I also did. But this would change the behavior of skia sharp. I would love to have the fontmanager from windows but let free type render them

**Comment by soerendd** [comment]
Another idea would be to bypass skia completely and do all the text related stuff in a different library utilizing free type and maybe also harfbuzz

**Comment by Gillibald** [comment]
Freetype doesn't require FontConfig that's just the default implementation that is being used. There is always a dictionary based font manager.

**Comment by soerendd** [comment]
@Gillibald I know, I know. But for a end user of skiasharp it's not easy to build this dictionary. And tbh, as a avalonia user I do even care less about these things. That's why I propose to build a second infrastructure either with only building the dictionary or to handle text completely on its own to get away from implementation details of skia

**Comment by Gillibald** [comment]
For Avalonia we can use any font manager implementation we want. In the end we are always able to load a Typeface via stream. By default we use Skia's font manager. Long term we will introduce our own FontConfig binding to better customize the lookup routine for Linux systems.

Being able to choose the font host implementation (DirectWrite, Freetype, X) at runtime would give us the flexibility we need. The font manager implementation isn't important.

**Comment by Ziriax** [comment]
Yes, I would love to see a version of SkiaSharp where you can inject the font manager and font rendering system at runtime. This won't be easy, since I think Skia is not design that way.

In an experiment I hacked this together to allow setting it once at program startup, before any SkiaSharp call was done. That was easy, but not really usable?

**Comment by HinTak** [comment]
Fwiw, in skia-python, we let you use 
`SkFontMgr_New_Custom_Empty` explicitly as an alternative to the default: 
https://github.com/kyamagu/skia-python/blob/e0b030c14e33f70880cfcc502eaeed898a77fc3c/src/skia/Font.cpp#L709
(Of course, on Linux the default is also freetype).
